betty/job/executor/__init__.py

```diff
@@ -0,0 +1,184 @@
+"""
+Schedule jobs using an executor.
+
+This module is internal.
+"""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from asyncio import run, gather, get_running_loop, Future, as_completed
+from dataclasses import dataclass
+from os import cpu_count
+from queue import Empty
+from typing import (
+    MutableSequence,
+    Callable,
+    Awaitable,
+    ParamSpec,
+    TypeVar,
+    TYPE_CHECKING,
+    Generic,
+    Any,
+    TypeAlias,
+)
+
+from typing_extensions import override
+
+from betty.job import Scheduler, Context, Job, JobEntry
+
+if TYPE_CHECKING:
+    from queue import Queue
+    from concurrent.futures import Executor
+    from threading import Event
+    from betty.concurrent import Lock, Semaphore
+
+_ContextT = TypeVar("_ContextT", bound=Context)
+_JobP = ParamSpec("_JobP")
+
+
+QueueItem: TypeAlias = tuple[Scheduler[Context], JobEntry]
+
+
+@dataclass
+class SharedState(Generic[_ContextT]):
+    """
+    A scheduler's state, shared with all its batched schedulers.
+    """
+
+    context: _ContextT
+    lock: Lock
+    queues: MutableSequence[Queue[QueueItem]]
+    semaphore: Semaphore
+    cancel: Event
+    finish: Event
+    job_count = 0
+
+
+class ExecutorScheduler(Generic[_ContextT], Scheduler[_ContextT]):
+    """
+    Schedule jobs using an executor.
+    """
+
+    def __init__(
+        self,
+        context_or_state: _ContextT | SharedState[_ContextT],
+        concurrency: int | None = None,
+    ):
+        if isinstance(context_or_state, SharedState):
+            self._state = context_or_state
+        else:
+            self._state = self._new_state(context_or_state)
+        self._queue = self._new_queue()
+        # @todo Is this appendage correct?
```

Comment:
It is! For jobs we basically work with a data structure that is a sequence of batches (like the result of an unflattened topological sort). A batch may be processed concurrently if there is no non-empty batch before it.

Reply:
This also means we need a way to indicate whether a queue if finished or not.

Reply:
Does that mean we can remove batches from the job API and reduce to just a single queue that is flushed between extension batches?